### ISSUE TITLE
The API does not take into account max_new_tokens

### ISSUE THREAD

**Issue by Tayyab-H** [oobabooga/text-generation-webui#1404]
### Bug Description

Im connecting to the oogabooga api and generating text however it does not obey the max_new_tokens parameter. I've set it to 85 and it continually generates prompts that are 200 tokens long. The web UI works fine and obeys the token length parameter however the api does not seem to do the same. I'll be honest the API could do with alot of work (I'd be happy to help at some point). It lacks quite a bit of functionality and docs. I am using the llama-7b-4bit-128g for the purposes of reproduction.

### Reproduction

```
data = { "prompt": BasePrompt+"\n"+ prompt + "|<ENDOFTURN>|" ,'encoder_repetition_penalty': 1.15, 'max_new_tokens': 85, 'temperature': 0.44, 'top_p': 1, 'use_world_info': True, 'use_memory': True, 'use_story': False, 'quiet': False, 'singleline':False}
response = requests.post('http://127.0.0.1:5000/api/v1/generate', headers=headers, data=json.dumps(data))
response = response.json()
resp = response["results"][0]["text"]
print(resp)

### System Info

```shell
i7-8700k
RTX 2080
16gb RAM
Samsung evo 980 SSD
Windows 10
```

**Comment by Reezlaw** [comment]
Have you tried _the other_ API? I ended up disabling the API extension and using the same port as the web interface (7850 by default) with the /run/textgen endpoint

**Comment by Tayyab-H** [comment]
im sorry THERES ANOTHER ENDPOINT? Any docs on this?

**Comment by AlTheGrath** [comment]
To get the API working you need to use http://{address}:{Port}/run/textgen (http://127.0.0.1:7860/run/textgen)  to work with the Gradio API.  
Example link: https://github.com/oobabooga/text-generation-webui/blob/main/api-example.py

you don't want to use the API flag but instead use --listen and --no-stream to get the API working essentially.

adding to what  @Reezlaw said, you need go into the "webui.py" and remove --chat and add --no-stream & --listen; additionally you can add share if you want to use the api from another computer 
--listen  is for the API itself 
--no-stream is needed for compatibility as the API needs to just return all the text at once rather than have it trickle in as it generates it. play around with this option in the webui and it will at clarity to its importance. 

in short,  on line 146 were you see 

run_cmd("python server.py --auto-devices --chat --model-menu")

replace with

run_cmd("python server.py --auto-devices --listen --no-stream --share --model-menu")

alternatively you could just change all of these setting in the "interface mode" tab in the UI every time you fire it up.

**Comment by CyberTimon** [comment]
Also, custom_stopping_strings don't seem to work either.
params['custom_stopping_strings'] = ["\n"]
Does nothing

**Comment by CyberTimon** [comment]
> To get the API working you need to use http://{address}:{Port}/run/textgen (http://127.0.0.1:7860/run/textgen) to work with the Gradio API. Example link: https://github.com/oobabooga/text-generation-webui/blob/main/api-example.py
> 
> you don't want to use the API flag but instead use --listen and --no-stream to get the API working essentially.
> 
> adding to what @Reezlaw said, you need go into the "webui.py" and remove --chat and add --no-stream & --listen; additionally you can add share if you want to use the api from another computer --listen is for the API itself --no-stream is needed for compatibility as the API needs to just return all the text at once rather than have it trickle in as it generates it. play around with this option in the webui and it will at clarity to its importance.
> 
> in short, on line 146 were you see
> 
> run_cmd("python server.py --auto-devices --chat --model-menu")
> 
> replace with
> 
> run_cmd("python server.py --auto-devices --listen --no-stream --share --model-menu")
> 
> alternatively you could just change all of these setting in the "interface mode" tab in the UI every time you fire it up.

Is there an alternative, to get this to work with streaming enabled?

**Comment by AlTheGrath** [comment]
I would write this in the parameters. 

instead of this

'stopping_strings': [],

I put 

 'stopping_strings': ["\n"],

something to note also is that this rollout seems heavily built to work around Vicuna so Id make your stopping string "Human:" as it seems Vicuna is designed around all prompts being split into  ### Human and ### Assistant.  You'll find that the API still returns the stopping strings so its good to remove it if its there. Here's the quick script I used for that. 

`def remove_delimiter(text, delimiter):
    """
    Removes a delimiter from a given string, if it exists in the string.
    """
    if delimiter in text:
        return text.replace(delimiter, "")
    else:
        return text`

"Is there an alternative, to get this to work with streaming enabled?"

not to my knowledge, I'd consider us luckily the API works at all its been on the frits for the last couple of rollouts due to a parameter change.

**Comment by danpizzy** [comment]
> 

i dont understand how this is done im sorry im still a noob to this stuffim using the vicuna model and want it to stop output after the response and not use###human

**Comment by f-ale** [comment]
I'm also observing the same issue on the /api/v1/generate endpoint.

**Comment by xynta** [comment]
Try to use "max_tokens"
